README.rst
.. image:: https://github.com/morepath/more.cors/workflows/CI/badge.svg?branch=master
   :target: https://github.com/morepath/more.cors/actions?workflow=CI
   :alt: CI Status

.. image:: https://coveralls.io/repos/github/morepath/more.cors/badge.svg?branch=master
    :target: https://coveralls.io/github/morepath/more.cors?branch=master

.. image:: https://img.shields.io/pypi/v/more.cors.svg
  :target: https://pypi.org/project/more.cors/

.. image:: https://img.shields.io/pypi/pyversions/more.cors.svg
  :target: https://pypi.org/project/more.cors/



more.cors: CORS support for Morepath
====================================

This package adds CORS support to Morepath.


Quick start
-----------

Install ``more.cors``:

.. code-block:: console

    $ pip install -U more.cors

Extend your App class from CORSApp:

.. code-block:: python

    from more.cors import CORSApp


    class App(CORSApp):
        pass

This will add basic CORS support to your Morepath app.


Settings
--------

more.cors provides settings in the 'cors' section. Here are the defaults:

.. code-block:: python

    @App.setting_section(section='cors')
    def cors_settings():
        return {
            'allowed_verbs': ['GET', 'PUT', 'POST', 'PATCH', 'DELETE', 'OPTIONS'],
            'allowed_origin': '*',
            'expose_headers': ['Content-Type', 'Authorization'],
            'allowed_headers': ['Content-Type', 'Authorization'],
            'max_age': 60,
            'allow_credentials': False
        }

The following settings are available:

allowed_verbs
  A list of allowed HTTP request methods.

allowed_origin
  A URI that may access the resource.
  For requests **without credentials**, "*" can be used as a wildcard,
  allowing any origin to access the resource.

expose_headers
  A list of HTTP headers which can be exposed as part of the response.

allowed_headers
  A list of HTTP headers which can be used during the actual request.

max_age
  Maximum number of seconds the results of a preflight request can be cached.

allow_credentials
  Boolean which indicates whether or not the actual request can be made using
  credentials.
  Credentials are cookies, authorization headers or TLS client certificates.


Specify CORS settings for a single view
---------------------------------------

more.cors exposes the ``App.cors()`` class method.
This can be used to specify settings for a single view:

.. code-block:: python

    App.cors(
        model=Root,
        view_name='view2',
        allowed_headers=['Cache-Control'],
        expose_headers=['Cookie'],
        allowed_origin='http://foo.com',
        allow_credentials=True,
        max_age=10
    )

model
  Specifies the corresponding view model.

view_name
  Is needed when you use a named view.

allowed_headers, expose_headers, allowed_origin, allow_credentials, max_age
  The settings which can be specified. For details see Settings_.


CHANGES.rst
CHANGES
=======

0.4 (unreleased)
----------------

- Fix Flake8.

- Drop support for Python 3.4 and 3.5.

- Add support for Python 3.9.

- Use GitHub Actions for CI.


0.3 (2020-04-26)
----------------

- **Removed**: Removed support for Python 2.
  
  You have to upgrade to Python 3 if you want to use this version.

- Added support for Python 3.7 and 3.8 and PyPy 3.6.

- Make Python 3.7 the default testing environment.

- Add integration for the Black code formatter.


0.2 (2018-06-11)
----------------

- Fix list of allowed verbs for failed (unauthorized) preflight (see `PR #4`_).

.. _PR #4: https://github.com/morepath/more.cors/pull/4


0.1 (2018-06-08)
----------------

- Initial public release


more/cors/tests/test_cors.py
from webtest import TestApp as Client
from webob.exc import HTTPUnauthorized
import morepath
from more.cors import CORSApp


class App(CORSApp):
    pass


class Object:
    pass


class FailedObject:
    pass


class UnauthorizedObject:
    pass


class Root:
    pass


@App.path(path='/', model=Root)
def get_root(request):
    return Root()


@App.path(path='obj', model=Object)
def get_object(request):
    return Object()


@App.path(path='failedobj', model=FailedObject)
def get_failed_object(request):
    raise Exception()


@App.path(path='unauthorized', model=UnauthorizedObject)
def get_unauthorized_object(request):
    raise HTTPUnauthorized


@App.json(model=Root)
def view(context, request):
    return {'view': 'index'}


@App.json(model=Root, request_method='POST')
def post_view(context, request):
    return {'view': 'index-post'}


@App.json(model=Root, name='view2')
def view2(context, request):
    return {'view': 'view2'}


@App.json(model=Root, name='view2', request_method='POST')
def post_view2(context, request):
    return {'view': 'view2-post'}


App.cors(
    model=Root,
    view_name='view2',
    allowed_headers=['Cache-Control'],
    expose_headers=['Cookie'],
    allowed_origin='http://localhost.localdomain',
    allow_credentials=True,
    max_age=10
)


def get_client(app):
    class Tapp(app):
        pass
    morepath.autoscan()
    morepath.commit(Tapp)
    c = Client(Tapp())
    return c


def test_cors_preflight():
    c = get_client(App)

    r = c.options('/', headers={'Origin': 'http://hello.world.com/'})

    assert r.headers.get(
        'Access-Control-Allow-Methods').split(',') == ['OPTIONS',
                                                       'GET',
                                                       'POST']
    assert r.headers.get('Access-Control-Allow-Origin') == '*'
    assert r.headers.get(
        'Access-Control-Allow-Headers').split(',') == ['Content-Type',
                                                       'Authorization']
    assert r.headers.get(
        'Access-Control-Expose-Headers').split(',') == ['Content-Type',
                                                        'Authorization']

    assert r.headers.get('Access-Control-Allow-Credentials') is None
    assert r.headers.get('Access-Control-Max-Age') == '60'


def test_cors_override():
    c = get_client(App)
    r = c.options('/view2', headers={'Origin': 'http://hello.world.com/'})

    assert r.headers.get(
        'Access-Control-Allow-Methods').split(',') == [
            'OPTIONS', 'GET', 'POST']
    assert r.headers.get(
        'Access-Control-Allow-Origin') == 'http://localhost.localdomain'
    assert r.headers.get(
        'Access-Control-Allow-Headers').split(',') == ['Cache-Control']
    assert r.headers.get(
        'Access-Control-Expose-Headers').split(',') == ['Cookie']

    assert r.headers.get('Access-Control-Allow-Credentials') == 'true'
    assert r.headers.get('Access-Control-Max-Age') == '10'


def test_cors_non_preflight():
    c = get_client(App)

    r = c.get('/', headers={'Origin': 'http://hello.world.com/'})

    assert r.headers.get('Access-Control-Allow-Origin') == '*'
    assert r.headers.get(
        'Access-Control-Expose-Headers').split(',') == ['Content-Type',
                                                        'Authorization']

    assert r.headers.get('Access-Control-Allow-Credentials') is None
    assert r.headers.get('Access-Control-Max-Age') is None


def test_cors_unauthorized_preflight():
    c = get_client(App)

    r = c.options('/unauthorized')

    assert r.headers.get(
        'Access-Control-Allow-Methods').split(',') \
        == ['OPTIONS'] + c.app.settings.cors.allowed_verbs


def test_cors_no_allowed_verbs():
    @App.setting(section='cors', name='allowed_verbs')
    def get_allowed_verbs():
        return []

    c = get_client(App)

    c.options(
        '/',
        headers={'Origin': 'http://hello.world.com/'},
        status=404
    )


def test_cors_failed_view():

    c = get_client(App)

    r = c.options('/failedobj', expect_errors=True)

    assert r.status_code == 404


more/cors/action.py
import dectate
import reg
from morepath.directive import PathCompositeAction


class CORSAction(dectate.Action):

    filter_convert = {
        'model': dectate.convert_dotted_name
    }
    app_class_arg = True

    def __init__(self, model, view_name=None,
                 allowed_origin=None,
                 allowed_headers=None,
                 expose_headers=None,
                 allow_credentials=None,
                 max_age=None):
        self.model = model
        self.view_name = view_name
        self.allowed_origin = allowed_origin
        self.allowed_headers = allowed_headers
        self.expose_headers = expose_headers
        self.allow_credentials = allow_credentials
        self.max_age = max_age

    def identifier(self, app_class):
        return str((app_class, self.model, self.view_name))

    def perform(self, obj, app_class):
        allowed_origin = self.allowed_origin
        if allowed_origin is not None:
            app_class.get_cors_allowed_origin.register(
                lambda self, model, request, requested_origin: allowed_origin,
                model=self.model, view_name=self.view_name)

        allowed_headers = self.allowed_headers
        if allowed_headers is not None:
            app_class.get_cors_allowed_headers.register(
                (lambda self, model, request,
                 requested_headers: allowed_headers),
                model=self.model, view_name=self.view_name
            )

        expose_headers = self.expose_headers
        if expose_headers is not None:
            app_class.get_cors_expose_headers.register(
                lambda self, model, request: expose_headers,
                model=self.model, view_name=self.view_name
            )

        allow_credentials = self.allow_credentials

        if allow_credentials is not None:
            app_class.get_cors_allow_credentials.register(
                lambda self, model, request: allow_credentials,
                model=self.model, view_name=self.view_name
            )

        max_age = self.max_age

        if max_age is not None:
            app_class.get_cors_max_age.register(
                lambda self, model, request: max_age,
                model=self.model, view_name=self.view_name
            )


more/cors/main.py
import morepath
from morepath.publish import resolve_model, get_view_name
from webob.exc import HTTPUnauthorized
import dectate
import reg
from . import action


class App(morepath.App):

    _cors = dectate.directive(action.CORSAction)

    @classmethod
    def cors(cls, *args, **kwargs):
        cls._cors(*args, **kwargs)(None)

    @reg.dispatch_method(
        reg.match_instance('model'),
        reg.match_key('view_name',
                      lambda self, model, request,
                      requested_origin: request.view_name))
    def get_cors_allowed_origin(self, model, request, requested_origin):
        return self.settings.cors.allowed_origin

    @reg.dispatch_method(
        reg.match_instance('model'),
        reg.match_key('view_name',
                      lambda self, model, request,
                      requested_method: request.view_name))
    def get_cors_allowed_methods(self, model, request, requested_method):
        if model == None:
            return self.settings.cors.allowed_verbs
        res = []
        for m in self.settings.cors.allowed_verbs:
            f = self.get_view.by_predicates(
                model=model.__class__, name=request.view_name,
                request_method=m)
            if f and f.component:
                res.append(m)
        return res

    @reg.dispatch_method(
        reg.match_instance('model'),
        reg.match_key('view_name',
                      lambda self, model, request,
                      requested_headers: request.view_name))
    def get_cors_allowed_headers(self, model, request, requested_headers):
        return self.settings.cors.allowed_headers

    @reg.dispatch_method(
        reg.match_instance('model'),
        reg.match_key('view_name',
                      lambda self, model, request: request.view_name))
    def get_cors_expose_headers(self, model, request):
        return self.settings.cors.expose_headers

    @reg.dispatch_method(
        reg.match_instance('model'),
        reg.match_key('view_name',
                      lambda self, model, request: request.view_name))
    def get_cors_allow_credentials(self, model, request):
        return self.settings.cors.allow_credentials

    @reg.dispatch_method(
        reg.match_instance('model'),
        reg.match_key('view_name',
                      lambda self, model, request: request.view_name))
    def get_cors_max_age(self, model, request):
        return self.settings.cors.max_age


CORSApp = App


@App.setting_section(section='cors')
def cors_settings():
    return {
        'allowed_verbs': ['GET', 'PUT', 'POST', 'PATCH', 'DELETE', 'OPTIONS'],
        'allowed_origin': '*',
        'expose_headers': ['Content-Type', 'Authorization'],
        'allowed_headers': ['Content-Type', 'Authorization'],
        'max_age': 60,
        'allow_credentials': False
    }


@App.tween_factory(over=morepath.EXCVIEW)
def cors_tween(app, handler):
    def cors_handler(request):
        preflight = request.method == 'OPTIONS'
        if preflight:
            response = request.ResponseClass()
        else:
            response = handler(request)

        try:
            context = resolve_model(request) or app
        except HTTPUnauthorized:
            context = None
        except Exception as e:
            context = app

        _marker = []
        if getattr(request, 'view_name', _marker) is _marker:
            request.view_name = get_view_name(request.unconsumed)

        # Access-Control-Allow-Methods
        requested_method = request.headers.get('Access-Control-Request-Method')

        if preflight:
            allowed_methods = app.get_cors_allowed_methods(
                context, request, requested_method)
            if not allowed_methods:
                return request.ResponseClass(status_code=404)
            response.headers.add('Access-Control-Allow-Methods',
                                 ','.join(['OPTIONS'] + allowed_methods))

        # Access-Control-Allow-Headers
        requested_headers = request.headers.get(
            'Access-Control-Request-Headers')

        if preflight:
            allowed_headers = app.get_cors_allowed_headers(
                context, request, requested_headers)
            response.headers.add(
                'Access-Control-Allow-Headers', ','.join(allowed_headers))

        # Access-Control-Allow-Origin
        requested_origin = request.headers.get('Origin')

        if requested_origin or preflight:
            allowed_origin = app.get_cors_allowed_origin(
                context, request, requested_origin)
            response.headers.add('Access-Control-Allow-Origin', allowed_origin)
            if allowed_origin and allowed_origin != '*':
                response.headers.add('Vary', 'Origin')

        # Access-Control-Expose-Headers
        exposed_headers = app.get_cors_expose_headers(context, request)
        response.headers.add('Access-Control-Expose-Headers',
                             ','.join(exposed_headers))

        # Access-Control-Allow-Credentials
        allow_credentials = app.get_cors_allow_credentials(context, request)
        if allow_credentials is True:
            response.headers.add('Access-Control-Allow-Credentials', 'true')

        if preflight:
            max_age = app.get_cors_max_age(context, request)
            response.headers.add('Access-Control-Max-Age',
                                 str(max_age))

        return response
    return cors_handler
